### PR TITLE
05/27 Add more site content, Slideshow image display, and Menu bar

### PR DESCRIPTION
**Pull request by nif33**
Screenshot of current UI attached:
image

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -8,39 +8,72 @@
   </head>
   <body>
     <div id="content">
+      <div id="menu">
+        <h2>
+        <a href="index.html">Home</a>&nbsp&nbsp
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/index.html:13]: Don't know much about &nbsp but found about this in a style guide.

> If you need to use non-breaking spaces, prefer style="white-space: nowrap" on the container tag over using the &nbsp; character entity where possible.

This is probably saying that this could be solved with CSS.

https://www.w3schools.com/cssref/pr_text_white-space.asp

File: portfolio/src/main/webapp/index.html [hunk 2]
```diff
@@ -8,39 +8,72 @@
...
         <h1>Jennifer Dai's Portfolio</h1>
-        <a href="https://ca.linkedin.com/in/jennifer-dai">LinkedIn</a>
-        <a href="https://github.com/nif33">Github</a>
       </div>
 
       <p>Click here to get a random fact about me:</p>
-      <button onclick="addRandomFact()">Tell me more!</button>
-      <div id="fact-container"></div>
+        <button onclick="addRandomFact()">Tell me more!</button>
+        <marquee><div id="fact-container"></div></marquee>
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/index.html]: Bad indentation?

Comment (juanalonsocruz) [inline portfolio/src/main/webapp/index.html]: Marquee seems to be obsolete. I don't mind it that you use it since this is just a starter project but if you are interested more in animations, you could check out [CSS Animations](https://www.w3.org/TR/css-animations-1/)

https://stackoverflow.com/questions/31951282/why-is-marquee-deprecated-and-what-is-the-best-alternative

File: portfolio/src/main/webapp/index.html [hunk 3]
```diff
@@ -8,39 +8,72 @@
...
+        This is my dog: William. People always tell me his name is too human and I know, but I did not choose his name. 
+        He is a 7 year old Miniaature Poodle. His hobbies are taking naps, eating snacks, and barking at firetrucks.
+      </p>
+      <div class="slideshow">
         <img class="dog-slides" src="/images/dog/dog0.jpg">
         <img class="dog-slides" src="/images/dog/dog1.jpg">
         <img class="dog-slides" src="/images/dog/dog2.jpg">
+        
+        <button onclick="nextSlide(0, -1)">Previous</button>
+        <button onclick="nextSlide(0, 1)">Next</button>
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/index.html]: Having JS looking calls in HTML seems a little bit messy. The thing that bothers me the most is explicit arguments for the function in HTML. Not too familiar with this but maybe have a function called nextDogSlide(), prevDogSlide(), nextPlaceSlide(), and prevPlaceSlide() and in their definitions call another function that actually does the work with the specific params. Would like to see what Andrew thinks about this.

On the same topic of JS functions calls in HTML, maybe you should be doing this fully in JS. This is what I found online
https://stackoverflow.com/questions/5871640/why-is-using-onclick-in-html-a-bad-practice
https://www.reddit.com/r/javascript/comments/3kamqn/why_is_inline_onclickfunction_considered_such_a/

So a solution can be something like what is in the stackoverflow link or something like

HTML:
`<button id="next-dog-slide-button">Previous</button>`

JS:
```
document.getElementById('next-dog-slide-button').onclick = function() {
  // Write how to handle this click here
}
```

Again, I would like Andrew's take on this.

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -17,12 +17,45 @@
  */
 function addRandomFact() {
   const facts =
-      ['I have a pet dog', 'I like to code'];
+      ['I have a pet dog!', 'I like to code!', 'Marquees are my favourite HTML element!'];
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js:20]: Uh-oh. Now I feel bad about my marquees comment!

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -17,12 +17,45 @@
...
   const factContainer = document.getElementById('fact-container');
   factContainer.innerText = fact;
-}
\ No newline at end of file
+}
+
+/**
+ * Controls slideshow image display on the page
+ */
+var slideIndices = [0, 0, 0]; // array of slide indices for each slide deck
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js]: https://google.github.io/styleguide/jsguide.html#features-use-const-and-let

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -17,12 +17,45 @@
...
+for(var i = 0; i < slideIndices.length; i++) {
+    showSlides(i, slideIndices[i]);
+}
+
+function nextSlide(deckNum, direction) {
+  showSlides(deckNum, slideIndices[deckNum] += direction);
+}
+
+// Returns corresponding class name for given deckNum
+function deckNumToName(deckNum) {
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js]: Its ok to use strings instead of numbers. So need to have this method.

Also, If I ever see in the wild nextSlide(1, 2), it is hard to tell what this function does without looking at its definition. It would be much easier to read if I saw nextSlide(PLACES_DECK, NEXT_SLIDE) where PLACES_DECK and NEXT_SLIDE are constants defined at the top of the file. PLACES_DECK could be defined as the string "dog-slides" and NEXT_SLIDE could be the integer 1 (since you are making this work using arithmetic). For the string constant, you might wonder why not just hard code the string. This is just convention that if you have the same string multiple times in the file might as well store it in a constant cause then it would be a lot easier to modify (if it ever needs to be) since you would only have to change it in one place instead of in multiple places in the file.

If something doesn't make sense, just add to this comment with questions.

File: portfolio/src/main/webapp/script.js [hunk 7]
```diff
@@ -17,12 +17,45 @@
...
+  if(deckNum === 0)
+    return "dog-slides"
+  else if(deckNum == 1)
+    return "place-slides"
+  else
+    return "food-slides"
+}
+
+function showSlides(deckNum, slideIndex) {
+  var slides = document.getElementsByClassName(deckNumToName(deckNum));
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js]: What do you think of only having one img element in the html and setting/changing its src attribute in js instead of having N img tags in the HTML and setting display to none?

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -17,12 +17,45 @@
...
+
+  slideIndex = Math.abs(slideIndex % slides.length) // wrap around index
+
+  for (var i = 0; i < slides.length; i++) { // display none except current index
+      slides[i].style.display = "none";  
+  }
+  slides[slideIndex].style.display = "block";
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js]: Display stuff might be better suited to do in CSS.

**Review by juanalonsocruz — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/index.html (hunk 2 above)
Comment (nif33) [reply portfolio/src/main/webapp/index.html]: I know they're obsolete, I just think they're hilarious to see these days 😆  a blast from the past.

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (nif33) [reply portfolio/src/main/webapp/index.html:13]: I'm trying to use the whitespace here to add spaces between text and not for controlling whether the text wraps. white-space: nowrap would not work for that since it collapses spaces, can I get a link to the style guide to see the other options?
https://www.w3schools.com/cssref/pr_text_white-space.asp indicates an option using CSS is to use pre-wrap and the space ' '  character rather than &nbsp. Would that be better? I used &nbsp because then you can easily see where the spaces are.

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (ajpaon) [reply portfolio/src/main/webapp/index.html]: One term you'll hear sometimes is "[code smell](https://en.wikipedia.org/wiki/Code_smell)." I would say Juan is picking up on a code smell here.

So why is it "smelly" to have these function calls with explicit arguments in HTML? I can think of a few reasons:

- It's not obvious at a glance what the different arguments mean.

- It's hard to catch typos; since these function calls are just strings in HTML your editor won't highlight syntax errors

- Most important, in my opinion: it's hard to modify. Imagine you want to add a new slide show between dog photos and place photos. After you add your new slideshow div you need to remember to update the function arguments for the place photos and food photos (because now they are the 3rd and 4th slideshows instead of 2nd and 3rd.)

Overall: this solution works, but I think we can design a solution that (1) works just as well and (2) is easier to reason about and modify. I think Juan's suggestions about moving more logic to JS are a good starting point.

File: portfolio/src/main/webapp/index.html [hunk 9]
```diff
@@ -8,39 +8,72 @@
...
         <img class="place-slides" src="/images/place/place1.jpg">
         <img class="place-slides" src="/images/place/place2.jpg">
+
+        <button onclick="nextSlide(1, -1)">Previous</button>
+        <button onclick="nextSlide(1, 1)">Next</button>
       </div>
 
+      <h2>My Favourite Food</h2>
+      <p>
+        My favourite food is any brunch food because there are no rules at bunch, anything can be a brunch food if it tries.
```
Comment (ajpaon) [inline portfolio/src/main/webapp/index.html]: so true!
btw there is a typo (bunch -> brunch)

File: portfolio/src/main/webapp/script.js (hunk 6 above)
Comment (ajpaon) [reply portfolio/src/main/webapp/script.js]: This could be a good case to use an "enum".

JS doesn't have an enum keyword, but the style guide has a commonly-used way to work around that. You can ignore the part talking about @enum annotations. https://google.github.io/styleguide/jsguide.html#features-objects-enums

**Review by ajpaon — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 7 above)
Comment (nif33) [reply portfolio/src/main/webapp/script.js]: I could definitely change it to this, I think it would be more readable. I had the images in the HTML before because I didn't have the slideshow.
A tradeoff would be that if the script doesn't load for whatever reason, there would only be one image though.

Comment (nif33) [reply portfolio/src/main/webapp/script.js]: I think that would be more readable. The trade-off would be that if the script doesn't load for any reason there would only be one image but I suppose that doesn't happen very often.

Comment (nif33) [reply portfolio/src/main/webapp/script.js]: I have changed this but I am noticing that the images take longer to load the first time they are viewed and I am unsure how to handle this issue while keeping all the images in the javascript?

File: portfolio/src/main/webapp/index.html [hunk 10]
```diff
@@ -21,8 +21,8 @@ <h1>Jennifer Dai's Portfolio</h1>
       </div>
 
       <p>Click here to get a random fact about me:</p>
-        <button onclick="addRandomFact()">Tell me more!</button>
-        <marquee><div id="fact-container"></div></marquee>
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/index.html]: Sorry! You can keep it if you want to! Not a big deal here.

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -30,32 +30,57 @@ function addRandomFact() {
...
+function changeSlide(deckNum, direction) {
+  const deckSize = SLIDES[deckNum].length;
+  let slideIndex = slideIndices[deckNum];
+  slideIndex += direction;
+  if(slideIndex === deckSize) { // wrap around
+      slideIndex = 0;
+  }
+  else if(slideIndex === -1) {
+      slideIndex = deckSize - 1;
+  }
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js]: Why did you change from modular math? I think what you had before was cleaner. If you have a reason, then you don't have to switch back.

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -30,32 +30,57 @@ function addRandomFact() {
...
+  slideIndices[deckNum] = slideIndex;
 }
 
-function nextSlide(deckNum, direction) {
-  showSlides(deckNum, slideIndices[deckNum] += direction);
+document.getElementById("prev-dog-slide-button").onclick = function() {
+  changeSlide(decks.DOG_DECK, -1);
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js]: Just by reading this line, it is not cleared what the -1 does. It can be ambiguous. I can guess what deck.DOG_DECK is used for.

Maybe change have on top of this file something like`const PREV_SLIDE_DIR = -1` and `const NEXT_SLIDE_DIR = 1` and then use that constant variable here.

File: portfolio/src/main/webapp/index.html (hunk 1 above)
Comment (juanalonsocruz) [reply portfolio/src/main/webapp/index.html:13]: Oh thank for the context! I think its fine to keep it with &nbsp.

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (nif33) [reply portfolio/src/main/webapp/script.js]: I realized that going in the backwards direction would go 2, 1, 0, 1 (because it is modded and then the absolute value is taken) so I would have to take the absolute value at a different place in the code for it to work which did not seem legible. If I figure out another way to do that math I'll change it back to modular math

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -17,12 +17,72 @@
...
+ */
+const decks = { // Enumerate slide decks
+  DOG_DECK: 0,
+  PLACE_DECK: 1,
+  FOOD_DECK: 2,
+};
+
+const SLIDES = [["/images/dog/dog0.jpg", "/images/dog/dog1.jpg", "/images/dog/dog2.jpg"],
+  ["/images/place/place0.jpg", "/images/place/place1.jpg", "/images/place/place2.jpg"],
+  ["/images/food/food0.jpg", "/images/food/food1.jpg", "/images/food/food2.jpg"]]
```
Comment (juanalonsocruz) [inline portfolio/src/main/webapp/script.js]: Since you are only using these object values as indices for SLIDES, you can just merge there two into one.

```
const decks = { // Enumerate slide decks
  DOG_DECK: ["/images/dog/dog0.jpg", "/images/dog/dog1.jpg", "/images/dog/dog2.jpg"],
  PLACE_DECK: ["/images/place/place0.jpg", "/images/place/place1.jpg", "/images/place/place2.jpg"],
  FOOD_DECK: ["/images/food/food0.jpg", "/images/food/food1.jpg", "/images/food/food2.jpg"],
};

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 14]
```diff
@@ -27,6 +27,6 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    response.getWriter().println("<h1>Hello Jen!</h1>");
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Make sure this change is not part of this PR. Make is part of the next PR. To do this you would have to branch out.

**Comment by juanalonsocruz** [comment]
For some reason can't comment in a previous comment so doing it here.

For math comment:
Oh! The modulo issue is that in JS % is remainder and not modulo. Learning something new.

For img loading slow comment:
You can fix this by preloading the images in JS (instead of in html)
https://stackoverflow.com/questions/3646036/preloading-images-with-javascript

**Review by juanalonsocruz — APPROVED** [review]
Approving but please make changes to my comments.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 14 above)
Comment (juanalonsocruz) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Or since this PR is about to merged soon, just remove this change, make other fixes, merge, then add this line.

File: portfolio/src/main/webapp/script.js (hunk 13 above)
Comment (juanalonsocruz) [reply portfolio/src/main/webapp/script.js]: Responding to your comment on chat

I see I thought object keys when they aren't known types like what you did, would automatically convert to some kind of enum int. What I found hard to read was not the enum but the 2D array. Hard to tell from the initializer what it is. So keep the enum but change the 2D array to be an object.

const slides = { 
  decks.DOG_DECK: ["/images/dog/dog0.jpg", "/images/dog/dog1.jpg", "/images/dog/dog2.jpg"],
  decks.PLACE_DECK: ["/images/place/place0.jpg", "/images/place/place1.jpg", "/images/place/place2.jpg"],
  decks.FOOD_DECK: ["/images/food/food0.jpg", "/images/food/food1.jpg", "/images/food/food2.jpg"],
};
